File: frontend/utils/fileRouter/fileAnalyzer.ts
/**
 * File Analyzer - Frontend utility for file type detection
 * Mirrors backend FileAnalyzer logic for client-side validation
 */

export interface FileAnalysis {
  filename: string;
  fileSize: number;
  fileSizeMB: number;
  fileExtension: string;
  fileType: 'pdf' | 'image' | 'docx' | 'pptx' | null;
  category: 'document' | 'image' | 'unknown';
  isSupported: boolean;
}

export interface FileCategory {
  extensions: string[];
  mimeTypes: string[];
  category: 'document' | 'image';
  requiresExtraction: boolean;
  supportsNative: boolean;
}

const FILE_CATEGORIES: Record<string, FileCategory> = {
  pdf: {
    extensions: ['.pdf'],
    mimeTypes: ['application/pdf'],
    category: 'document',
    requiresExtraction: true,
    supportsNative: true,
  },
  image: {
    extensions: ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp'],
    mimeTypes: ['image/jpeg', 'image/png', 'image/gif', 'image/bmp', 'image/webp'],
    category: 'image',
    requiresExtraction: false,
    supportsNative: true,
  },
  docx: {
    extensions: ['.doc', '.docx'],
    mimeTypes: [
      'application/msword',
      'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    ],
    category: 'document',
    requiresExtraction: true,
    supportsNative: false,
  },
  pptx: {
    extensions: ['.ppt', '.pptx'],
    mimeTypes: [
      'application/vnd.ms-powerpoint',
      'application/vnd.openxmlformats-officedocument.presentationml.presentation',
    ],
    category: 'document',
    requiresExtraction: true,
    supportsNative: false,
  },
};

/**
 * Analyze a file and return detailed metadata
 */
export function analyzeFile(file: File): FileAnalysis {
  const fileExtension = getFileExtension(file.name);
  const fileType = detectFileType(fileExtension, file.type);
  const fileInfo = fileType ? FILE_CATEGORIES[fileType] : null;

  return {
    filename: file.name,
    fileSize: file.size,
    fileSizeMB: parseFloat((file.size / (1024 * 1024)).toFixed(2)),
    fileExtension,
    fileType,
    category: fileInfo?.category || 'unknown',
    isSupported: fileType !== null,
  };
}

/**
 * Detect file type based on extension and MIME type
 */
function detectFileType(
  extension: string,
  mimeType: string
): 'pdf' | 'image' | 'docx' | 'pptx' | null {
  for (const [fileType, info] of Object.entries(FILE_CATEGORIES)) {
    if (info.extensions.includes(extension)) {
      return fileType as any;
    }
    if (info.mimeTypes.includes(mimeType)) {
      return fileType as any;
    }
  }
  return null;
}

/**
 * Get file extension from filename
 */
function getFileExtension(filename: string): string {
  const lastDot = filename.lastIndexOf('.');
  return lastDot !== -1 ? filename.slice(lastDot).toLowerCase() : '';
}

/**
 * Validate file size
 */
export function validateFileSize(
  fileSizeMB: number,
  maxSizeMB: number = 50
): { isValid: boolean; error?: string } {
  if (fileSizeMB > maxSizeMB) {
    return {
      isValid: false,
      error: `File too large (${fileSizeMB}MB). Maximum: ${maxSizeMB}MB`,
    };
  }
  return { isValid: true };
}

/**
 * Get file category (quick helper)
 */
export function getFileCategory(filename: string): 'document' | 'image' | 'unknown' {
  const extension = getFileExtension(filename);
  const fileType = detectFileType(extension, '');
  
  if (fileType && FILE_CATEGORIES[fileType]) {
    return FILE_CATEGORIES[fileType].category;
  }
  return 'unknown';
}

/**
 * Check if file is supported
 */
export function isFileSupported(filename: string): boolean {
  const extension = getFileExtension(filename);
  const fileType = detectFileType(extension, '');
  return fileType !== null;
}

/**
 * Get supported file extensions as array
 */
export function getSupportedExtensions(): string[] {
  const extensions: string[] = [];
  Object.values(FILE_CATEGORIES).forEach(category => {
    extensions.push(...category.extensions);
  });
  return extensions;
}

/**
 * Format file size for display
 */
export function formatFileSize(bytes: number): string {
  if (bytes === 0) return '0 Bytes';
  
  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(2))} ${sizes[i]}`;
}

File: backend/utils/file_router/file_analyzer.py
"""
File Analyzer - Deep file analysis and metadata extraction
Detects file types, validates formats, extracts metadata
"""

import os
import mimetypes
from typing import Dict, Any, Optional, Tuple
from pathlib import Path


class FileAnalyzer:
    """
    Analyzes uploaded files to determine type, size, and compatibility
    """
    
    # Supported file types and their categories
    FILE_CATEGORIES = {
        'pdf': {
            'extensions': ['.pdf'],
            'mimetypes': ['application/pdf'],
            'category': 'document',
            'requires_extraction': True,
            'supports_native': True,  # Can be read natively by some models
        },
        'image': {
            'extensions': ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp'],
            'mimetypes': ['image/jpeg', 'image/png', 'image/gif', 'image/bmp', 'image/webp'],
            'category': 'image',
            'requires_extraction': False,
            'supports_native': True,  # Vision models can read directly
        },
        'docx': {
            'extensions': ['.doc', '.docx'],
            'mimetypes': ['application/msword', 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'],
            'category': 'document',
            'requires_extraction': True,
            'supports_native': False,  # Must extract text
        },
        'pptx': {
            'extensions': ['.ppt', '.pptx'],
            'mimetypes': ['application/vnd.ms-powerpoint', 'application/vnd.openxmlformats-officedocument.presentationml.presentation'],
            'category': 'document',
            'requires_extraction': True,
            'supports_native': False,  # Must extract text
        },
    }
    
    @staticmethod
    def analyze_file(filename: str, file_size: int) -> Dict[str, Any]:
        """
        Analyze a file and return detailed metadata
        
        Args:
            filename: Name of the uploaded file
            file_size: Size of the file in bytes
            
        Returns:
            Dict containing file analysis results
        """
        file_ext = Path(filename).suffix.lower()
        mime_type, _ = mimetypes.guess_type(filename)
        
        # Detect file type
        file_type = FileAnalyzer._detect_file_type(file_ext, mime_type)
        file_info = FileAnalyzer.FILE_CATEGORIES.get(file_type, {})
        
        return {
            'filename': filename,
            'file_size': file_size,
            'file_size_mb': round(file_size / (1024 * 1024), 2),
            'file_extension': file_ext,
            'mime_type': mime_type,
            'file_type': file_type,
            'category': file_info.get('category', 'unknown'),
            'requires_extraction': file_info.get('requires_extraction', True),
            'supports_native': file_info.get('supports_native', False),
            'is_supported': file_type is not None,
        }
    
    @staticmethod
    def _detect_file_type(extension: str, mime_type: Optional[str]) -> Optional[str]:
        """
        Detect file type based on extension and mime type
        
        Returns:
            File type key (pdf, image, docx, pptx) or None if unsupported
        """
        for file_type, info in FileAnalyzer.FILE_CATEGORIES.items():
            if extension in info['extensions']:
                return file_type
            if mime_type and mime_type in info['mimetypes']:
                return file_type
        return None
    
    @staticmethod
    def validate_file(filename: str, file_size: int, max_size_mb: int = 50) -> Tuple[bool, Optional[str]]:
        """
        Validate if file meets basic requirements
        
        Args:
            filename: Name of the file
            file_size: Size in bytes
            max_size_mb: Maximum allowed size in MB
            
        Returns:
            Tuple of (is_valid, error_message)
        """
        analysis = FileAnalyzer.analyze_file(filename, file_size)
        
        # Check if file type is supported
        if not analysis['is_supported']:
            supported_types = []
            for info in FileAnalyzer.FILE_CATEGORIES.values():
                supported_types.extend(info['extensions'])
            return False, f"Unsupported file type. Supported: {', '.join(supported_types)}"
        
        # Check file size
        if analysis['file_size_mb'] > max_size_mb:
            return False, f"File too large ({analysis['file_size_mb']}MB). Max: {max_size_mb}MB"
        
        return True, None
    
    @staticmethod
    def get_file_category(filename: str) -> str:
        """
        Quick method to get file category (document, image, unknown)
        
        Args:
            filename: Name of the file
            
        Returns:
            Category string
        """
        file_ext = Path(filename).suffix.lower()
        mime_type, _ = mimetypes.guess_type(filename)
        file_type = FileAnalyzer._detect_file_type(file_ext, mime_type)
        
        if file_type:
            return FileAnalyzer.FILE_CATEGORIES[file_type]['category']
        return 'unknown'
